File: typescript/src/model/expressions/wdl-literal.ts
/** Base literal expression abstraction. */
import type { WdlExpressionComponentType } from './wdl-expression.js';
import type { WdlValueExpression } from './wdl-value-expression.js';

/** Base class for scalar literal expressions that carry a concrete value. */
export abstract class WdlLiteral<T> implements WdlValueExpression<T> {
  /** Creates a literal from its underlying runtime value. */
  public constructor(private value?: T) {}

  /** Returns the literal value. */
  public getValue(): T | undefined {
    return this.value;
  }

  /** Sets the literal value. */
  public setValue(value: T | undefined): void {
    this.value = value;
  }

  /** Returns the broad expression family for this node. */
  public abstract componentType(): WdlExpressionComponentType;
}

File: python/src/wdl_model/model/expressions/wdl_literal.py
"""Generic literal expression nodes."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, Optional, TypeVar

from .wdl_value_expression import WdlValueExpression

T = TypeVar("T")


@dataclass
class WdlLiteral(WdlValueExpression[T], Generic[T]):
    """Literal expression carrying a concrete scalar value."""

    value: Optional[T] = None

    def getValue(self) -> Optional[T]:
        """Return the literal value."""
        return self.value

    def setValue(self, value: Optional[T]) -> None:
        """Set the literal value."""
        self.value = value

    def __str__(self) -> str:
        return str(self.value)
